zdm/repeat_grid.py

```diff
@@ -0,0 +1,1365 @@
+        
+        self.bmethod=bmethod
+        
+        # set these on init, to remember for update purposes
+        self.Exact = Exact
+        self.MC = MC
+        
+        # handles plotting
+        if opdir is not None:
+            #for plotting
+            import os
+            if not os.path.exists(opdir):
+                os.mkdir(opdir)
+            self.opdir=opdir
+            doplots=True
+        else:
+            self.opdir=None
+            doplots=False
+        
+        # redshift array
+        self.zvals=self.grid.zvals
+        
+        # checks we have the necessary data to construct Nfields and Tfield
+        if Nfields is not None:
+            self.Nfields=Nfields
+        elif grid.survey.meta.has_key('Nfields'):
+            self.Nfields = grid.survey.meta['Nfields']
+        else:
+            self.Nfields = 1
+            #raise ValueError("Nfields not specified")
+        
+        if Tfield is not None:
+            self.Tfield=Tfield
+        elif grid.survey.meta.has_key('Tfield'):
+            self.Tfield = grid.survey.meta['Tfield']
+        else:
+            raise ValueError("Tfield not specified")
+        
+        # calculates constant Rc in front of dN/dR = Rc R^Rgamma
+        # this needs to be updated if any of the repeat parameters change
+        #self.Rc,self.NRtot=self.calc_constant(verbose=verbose)
+        #grid.state.rep.RC=self.Rc
+        # this now updated the parameters above
+        self.calc_constant(verbose=verbose)
+        
+        if verbose:
+            print("Calculated constant as ",self.Rc)
+        
+        # number of expected FRBs per volume in a given field
+        self.Nexp_field = self.grid.dV*self.Rc
+        # undoes the time-dilation effect, which is included in the FRB rate scaling
+        self.Nexp_field *= (1.+self.grid.zvals) # previously was reduced by this amount
+        self.Nexp = self.Nexp_field * self.Nfields
+        
+        # the below has units of Mpc^3 per zbin, and multiplied by p(DM|z) for each bin
+        # Note that grid.dV already has a (1+z) time dilation factor in it
+        # This is actually accounted for in the rate scaling of repeaters
+        # Here, we want the actual volume, hence must remove this factor
+        # recently removed this from sim_repeaters function
+        self.tvolume_grid = (self.grid.smear_grid.T * (self.grid.dV * (1. + self.grid.zvals))).T
+        #volume_grid *= Solid #accounts for solid angle viewed at that beam sensitivity
+        #self.volume_grid = volume_grid
+        
+        
+        # to remove alpha effect from grid.sfr if alpha_method==1
+        if self.grid.state.FRBdemo.alpha_method==1:
+            self.use_sfr = self.grid.source_function(self.grid.zvals,self.grid.state.FRBdemo.sfr_n)
+        else:
+            self.use_sfr = self.grid.sfr
+        
+        self.calc_Rthresh(Exact=Exact,MC=MC,doplots=doplots)
+
+    def update(self,Rmin = None,Rmax = None,Rgamma = None):
```

Comment:
define `Rmin, Rmax, Rgamma` in doc string

Reply:
Done